### ISSUE TITLE
How to perform an input transform within the model?

### ISSUE THREAD

**Issue by jduerholt** [pytorch/botorch#480]
Dear all,

I have a question regarding the transformers defined in `botorch/botorch/models/transforms/input.py`. How can I use them in combination with a model like SingleTaskGP, so that the input transform is happening within the model itself. 

I do not want to do the transformation in advance since I have to define an equality constraint on top of the untransformed inputs within optimize_acqf.

Best,

Johannes

**Comment by jduerholt** [comment]
Also in the documentation of the ModelListGP it is stated that models can have input transforms (https://github.com/pytorch/botorch/blob/master/botorch/models/model_list_gp_regression.py#L36). But I was so far not able to figure out how to use them.

Here is the excerpt of the documentation of ModelListGP:

`Args:
       *gp_models: An variable number of single-output BoTorch models.
                If models have input/output transforms, these are honored
                individually for each model.  
`

**Comment by Balandat** [comment]
Hi, thanks for raising this issue.

Unlike the OutcomeTransforms, we haven't really hooked the input transforms to the models yet, primarily b/c there hasn't been much demand for it. So the ModelListGP docstring is a little ahead of its time.

It's not too hard to implement though, you can do something like 

```
class InputTransformableSingleTaskGP(SingleTaskGP):
    
    def __init__(self, **kwargs):
        input_transform = kwargs.pop("input_transform", None)
        super().__init__(**kwargs)
        self.input_transform = input_transform
        
    def forward(self, X):
        if self.input_transform is not None:
            X = self.input_transform(X)
        return super().forward(X)
```

Example nb: [demo_input_transform.ipynb.txt](https://github.com/pytorch/botorch/files/4919826/demo_input_transform.ipynb.txt)

Let me put hooking this up for all models back on the agenda - there are some consideration for some of the more advanced models that make this less straightforward, but we can probably work those out.

**Comment by jduerholt** [comment]
Hi,

thank you very much for your quick response.

Concerning your solution, I have a few questions and remarks:

- Is it sufficient to put the input transform only into the `forward` function? Because in the case of the outcome_transformer `Y_train` is transformed also in the `__init__` function.
- I think it could also make sense to implement an input transformer which performs a standardization (zero mean, unit variance).
- Furthermore, it could be interesting to be able to specify in the input transformer to which columns it should be applied. For example in my case, I have a few feature that are one hot encoded, and I want to apply the transform only to the numeric ones.
- The cherry on the top would be a transformer especially dedicated to the ModelListGP, which filters those features out that are not relevant for a submodel. By the help of such a filter one could use not only different trainings sets for the single models, but also different subsets of features. This could be relevant for multiobjective optimizations, especially when optimizing phyical experiments.

What do you think about these ideas? Do they make sense? 

I think, I will start working on implementing a StandardScaler as input transformer and the possibility to define those columns to which the transform should be applied.

Best,

Johannes

**Comment by Balandat** [comment]
Hi Johannes, 

If you're looking for a more convenient interface, I recommend taking a look at Ax, in whcih we have a number of transforms implemented in (see https://ax.dev/api/modelbridge.html#transforms). 

> Is it sufficient to put the input transform only into the forward function? Because in the case of the outcome_transformer Y_train is transformed also in the __init__ function.

I haven't tested this much, but I believe so - `forward`'s `X` are essentially the input features, so they will always be transformed there. The reason `Y_train` is transformed in the `__init__` function is that we store the transformed training outcomes on the model and retrieve them from there. 

> I think it could also make sense to implement an input transformer which performs a standardization (zero mean, unit variance).

Yeah that could be useful, too

> Furthermore, it could be interesting to be able to specify in the input transformer to which columns it should be applied. For example in my case, I have a few feature that are one hot encoded, and I want to apply the transform only to the numeric ones.

That makes sense. Again, you should probably take a look at Ax, which does a lot of this stuff automatically. 

> The cherry on the top would be a transformer especially dedicated to the ModelListGP, which filters those features out that are not relevant for a submodel. By the help of such a filter one could use not only different trainings sets for the single models, but also different subsets of features. This could be relevant for multiobjective optimizations, especially when optimizing phyical experiments.

That seems like a very reasonable thing to do. Not sure if you're aware, but our latest release introduced a few novel algorithms for multi-objective optimization, see e.g. the tutorial: https://github.com/pytorch/botorch/blob/master/tutorials/multi_objective_bo.ipynb

**Comment by jduerholt** [comment]
Hi Max,

thanks again for your quick response. I had already a look on Ax but I think it is a bit too high-level for what we want to do. 

We are working on an optimization problem for improving physical experiments in a multi-objective setting. Not all features are relevant for all objectives, so we need this additional filter on the ModelListGP level. Furthermore an equality constraint has to defined on some of the untransformed features. We are also not able to perform any closed loop experimentation, instead we optimize with respect to the known datapoints and then communicate the proposed experimental setups to other people who perform the actual experiments. For this reason we also play with different acquisition functions and test them on portions of the already available data.

Furthermore not all regions of the pareto front are relevant, therefore we do not use the novel algorithms, you metnioned, instead we use a custom objective based on non-linear transformations of the single predictions which rewards only those solutions which are lying in the region of the pareto front which is of relevance. Due to the non-linear scalarization, we use MCAcquisitionFunctions.

From looking at Ax, I had the impression that it is better to implement this kind of workflow with low-level botorch. This was also the reason for asking for the input-transformers in botorch, and not using those implemented in Ax.

What do you think?

Best,

Johannes

**Comment by Balandat** [comment]
Hi Johannes, that makes sense, if you're in a non-standard setup and are using custom acquisition functions / optimization methods, then you'd want to implement those in botorch (btw, do you have a reference for the strategy you are using to encode the region of relevance of the Pareto frontier? - cc @sdaulton). 

Nevertheless, you can hook you custom methods into Ax in order to benefit from Ax's loop control, tracking, and db/storage capabilities, see e.g. at the end of [this tutorial](https://botorch.org/tutorials/custom_acquisition). We're also actively working on making using custom botorch modules in Ax easier, so stay tuned on that end.

**Comment by jduerholt** [comment]
Hi Max,

thanks for the link to the tutorial, it is really helpful. I think once we have setup and tested our customizations it makes sense to migrate them also to Ax. Equality on the feature level are already implemented in Ax?

Nevertheless, we will first start to implement the missing things on the side of the input transformers and the filtering in the ModelListGP, which could be helpful for multiobjective optimization in general. 

Concerning the strategy of encoding the region of relevance, we are trying to do something very simple (hopefully it will work ;)) borrowed from classical DOEs and which I think is already done in botorch for implementing constraint optimizations. I try to explain:

Assuming one has different outputs to optimize, I pass every of them to a custom sigmoid function which results in something between zero and one. The turning point of the sigmoid is chosen in a way that those predictions which are of interest get a reward of one and those which are not interesting of zero. This is especially useful when one wants to maximize an output against a given target value. Reaching values even higher than the target value is nice, but not really necessary. So one do not get any further reward, instead the optimizer can focus on the other outputs. The steepness of the custom sigmoid is chosen based on domain knowledge and the existing data.

The scalarized single objective is then computed as geometric mean of the outcomes of the sigmoids. Of course there are also other non-linear functions that one could choose for this purpose, but we started with sigmoids 😉 If one has a hierarchy of objectives (which we do not have in our case), one could also go for the chimera scalarization (https://pubs.rsc.org/en/content/articlelanding/2018/sc/c8sc02239a#!divAbstract), which could be also interesting for botorch (but so far I have not used it).

Best,

Johannes

**Comment by Balandat** [comment]
> Equality on the feature level are already implemented in Ax?

We don't have that implemented on the top level interface yet, but it should not be too hard to add. For the optimization we can easily pass the constraint down to the low-level optimizer, but we'll have to think a little more about how to generically generate random designs on some manifold / subspace if equality constraints present (cc @ldworkin). I botorch we currently support linear equality constraints - is that what you need for your application? Would there be a way to re-parameterize things and eliminate one of the parameters?

> The turning point of the sigmoid is chosen in a way that those predictions which are of interest get a reward of one and those which are not interesting of zero. This is especially useful when one wants to maximize an output against a given target value. Reaching values even higher than the target value is nice, but not really necessary. So one do not get any further reward, instead the optimizer can focus on the other outputs. 

This makes sense. I wonder how this compares to `qExpectedHypervolumeImprovement` with a properly chosen reference point - it seems that in your case you aren't necessarily interested in maximizing hypevolume, but achieving some target level for the different outputs. If this works out well for you we'd be happy to accept a PR for this.

> If one has a hierarchy of objectives (which we do not have in our case), one could also go for the chimera scalarization

Interesting, I am not familiar with this one, will take a look.

**Comment by jduerholt** [comment]
Yes, I saw that you provide in botorch linear equality constraints via scipy's implementation of SLSQP, and I wanted to use this, but therefor I need the input transformation happening within the model, because I cannot define the constraint on top of the standard transformed inputs (or at least I do not know how, and the solution with having the transform within the model is from my perspective also more elegant)

In principle the constraint is a D-dimensional (D>2) simplex constraint (all inputs sum up to one) on top a subset of the features. In the D=2 case, it is straight forward to eliminate one parameter, but I do not know how to do it in the higher dimensional case, especially when already the upper bounds for the D-n (n <D) parameters are chosen in a way that there is already for the D-n subset of paremeters the possibility that its sum exceeds 1 and by this violates the constraint. For this reason, I thought that it is the easiest way to handle the linear constraint via SLSQP.

If you have a better solution for this, I highly appreciate your feedback ;)

Concerning PRs, as soon as we have something useful (also for others) regarding the different things mentioned above, we'll try to file a pull request.

Best,

Johannes

**Comment by Balandat** [comment]
Interesting, I've run into a simplex constraint of the same nature a couple of times before. Might be worth thinking a bit more about this on our end.

**Comment by jduerholt** [comment]
Yeah, I think that could really make sense. Especially when optimizing the composition of mixtures which is of big relevance for example in the chemical industry, one runs almost always into this constraint. Perhaps, there is even a more elegant solution for it than using SLSQP. 

In the README.md (https://github.com/wujian16/Cornell-MOE/blob/master/README.md) of the CornellMOE package, it is mentioned that they support D-dimensional simplex constraints. Perhaps, it would be a good to start to have a look how they have implemented it.

Best,

Johannes

**Comment by jduerholt** [comment]
Hi Max,

I think I had misconception in reasoning when I was originally filing this issue. This morning, I did a bit of very simple math and I think it is already possible to use the equality constraint in botorch on top of the transformed (MinMax scaled or standard scaled) features. See the attached pdf document. Sorry for bothering you with something, that was already possible ...

Of course, this solution requires to write a bit of IO around the optimization loop, and hooking column specific input transformers directly into the model is the more elegant way. So presumably, we will implement the column specific StadardScaler into botorch and the filtering of features in the ModelListGP.

Best,

Johannes
[constraint.pdf](https://github.com/pytorch/botorch/files/4930227/constraint.pdf)
